Fix: restore generation of `sysWindowsResizeEvent` when borders are changed

SlySven opened the pull request:
This should close #6614.

Whilst looking through the code-base I was confused to see that the members `(int) mMainFrame{Top|Bottom|Right|Left}{Height|Width}` in the `TConsole` class were directly related to the `(int) mBorder{Top|Bottom|Left|Right}{Height|Width}` members. Also they are often manipulated as a set. I felt that it was constructive to refactor them into a common `std::tuple<int, int, int, int> mBorders`.

This refactoring meant it was also possible to detect when the borders were changed (in the new `(void) Host::setBorders(const std::tuple<int, int, int, int> borders)`) and to raise the missing `sysWindowResizeEvent` as needed.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: restore generation of `sysWindowsResizeEvent when boarders are changed

This should close #6614.

Whilst looking through the code-base I was confused to see that the
members `(int) mMainFrame{Top|Bottom|Right|Left}{Height|Width}` in the
`TConsole` class were directly related to the
`(int) mBorder{Top|Bottom|Left|Right}{Height|Width}` members. Also they are
often manipulated as a set. I felt that it was constructive to refactor
them into a common `std::tuple<int, int, int, int> mBorders`.

This refactoring meant it was also possible to detect when the borders
were changed (in the new
`(void) Host::setBorders(const std::tuple<int, int, int, int> borders)`)
and to raise the missing `sysWindowResizeEvent` as needed.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap reviewed (changes requested):
Mudlet crashes, after selecting a profile and clicking "Offline" button.
The profile popup vanishes, the black main window is shown, but adds no text, a few seconds later Mudlet vanishes completely. 
No helpful text is shown when started in CLI either. This is happening on Win.

vadi2 commented:
The bugfix is coming with too much refactoring and subsequently possibility for errors. Can we split the two so we just get the bugfix in 4.17, and the refactoring goes into 4.18? It would also be better to use a `struct` for this over unnamed fields in a tuple - a feature I really only properly learnt recently myself.

SlySven commented:
Ah, it was an issue during start-up where the `(TConsole*) Host::mpConsole`  member was still a `nullptr` in the new `(void) Host::setBorders(const std::tuple<int, int, int, int> borders)` which is being called from the `XMLimport` class code that loads the profile... I can see the fix - just building it to test.

SlySven committed:
Fix: stop crashes during startup

I forgot to check that a pointer was valid when a profile is being loaded,
which was causing fatal `nullptr` dereferences (Seg. Faults).

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> Can we split the two so we just get the bugfix in 4.17, and the refactoring goes into 4.18?

Not easily as I see it - it was the refactoring that enabled me to ensure the event was being generated whenever the border(s) were being changed - and the prior code that treated each of the four separately was not amenable (IMHO) to consistently monitor all four of them at the same time - and only generate a *single* event should *any* of them change.

> It would also be better to use a struct for this over unnamed fields in a tuple - a feature I really only properly learnt recently myself.

I get that - I've only just got to grips with the `std::tuple<T, ...>` template class myself. (I was toying with defining 4 `static const`s to use instead of `0` to `3`) - do you know of  any examples I could refer to for guidance on replacing this tuple with a `struct`?

vadi2 commented:
> Not easily as I see it 

OK.

Struct - have a look at `struct GameDetail` as an example!

Kebap commented on src/Host.cpp:
```diff
@@ -4164,3 +4160,20 @@ QPointer<TConsole> Host::parentTConsole(QObject* start) const
+void Host::setBorders(const std::tuple<int, int, int, int> borders)
+{
+    auto original = mBorders;
+    if (borders != original) {
+        mBorders = borders;
+        if (!mpConsole.isNull()) {
+            auto x = mpConsole->width();
+            auto y = mpConsole->height();
+            QSize s = QSize(x, y);
+            QResizeEvent event(s, s);
+            QApplication::sendEvent(mpConsole, &event);
+            mpConsole->raiseMudletSysWindowResizeEvent(x, y);
+        }
+    }
```
Anti-pattern: Don't indent the whole function by two levels for practically no rewards.
```suggestion
    if (borders == original) {
        return;
    }
    mBorders = borders;
    if (mpConsole.isNull()) {
        return;
    }
    auto x = mpConsole->width();
    auto y = mpConsole->height();
    QSize s = QSize(x, y);
    QResizeEvent event(s, s);
    QApplication::sendEvent(mpConsole, &event);
    mpConsole->raiseMudletSysWindowResizeEvent(x, y);
```

SlySven committed:
REvise: remove unneeded indentation by "failing out" early

Co-authored-by: Kebap <kebap_spam@gmx.net>

SlySven commented:
Guess what, Qt already has a (pair of) classes for this! See **[`QMargins`](https://doc.qt.io/qt-5/qmargins.html)** and **[`QMarginsF`](https://doc.qt.io/qt-5/qmarginsf.html)** :astonished:

vadi2 commented:
Oh! Even better.

SlySven committed:
Revise: use Qt own `QMargins` class as it is designed for what we want

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/TConsole.h:
```diff
@@ -240,10 +241,8 @@ class TConsole : public QWidget
-    int mMainFrameLeftWidth = 0;
-    int mMainFrameRightWidth = 0;
-    int mMainFrameTopHeight = 0;
+    // Order top, bottom, left, right:
```
```suggestion
```

SlySven commented on src/TConsole.cpp:
```diff
@@ -590,18 +583,20 @@ void TConsole::resizeEvent(QResizeEvent* event)
-        mpMainDisplay->resize(x - mMainFrameLeftWidth - mMainFrameRightWidth, y - mMainFrameTopHeight - mMainFrameBottomHeight - mpCommandLine->height());
+        x -= (mpLeftToolBar->width() + mpRightToolBar->width());
+        y -= mpTopToolBar->height();
+        // The mBorders will be all zeros for all but the MainConsole:
```
```suggestion
        // The mBorders components will be all zeros for all but the MainConsole:
```

SlySven committed:
Revise: remove one and tweak another comment

Kebap reviewed (approved):
Does not crash anymore and solves the original issue. Thanks!

vadi2 commented on src/TConsole.cpp:
```diff
@@ -83,36 +83,15 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
     ps->setContext(Qt::WidgetShortcut);
 
-    if (mType & CentralDebugConsole) {
+    if (mType == CentralDebugConsole) {
```
👍
